File: .github/scripts/lib/dedupe.mjs
import { mapCategory } from './category.mjs';

/**
 * Normalize a channel name for deduplication (spec §5.3, §4.3).
 * Preserve Unicode letters (\p{L}) and numbers (\p{N}) so Bengali / Hindi
 * names are not collapsed to empty strings.
 */
export function normalizeName(name) {
  return name
    .toLowerCase()
    .replace(/\b(hd|fhd|4k|tv)\b/g, '')
    .replace(/[^\p{L}\p{M}\p{N}\s-]/gu, '')
    .replace(/\s+/g, ' ')
    .trim();
}

async function sha256Hex16(input) {
  const buf = new TextEncoder().encode(input);
  const hash = await crypto.subtle.digest('SHA-256', buf);
  return Array.from(new Uint8Array(hash))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')
    .substring(0, 16);
}

/**
 * Deduplicate validated channels: one entry per normalized name, with ALL
 * passing URLs kept in `sources[]` sorted by latency (fastest first).
 * `streamUrl` is the fastest URL.
 *
 * @param {Array} validated  each has {name, url, latencyMs, language, tier, sourceId, category, logo}
 * @param {() => string} [nowIso]  injectable clock for tests
 * @returns {Promise<Array>} final ChannelFinal-shaped entries
 */
export async function pickBestRoutes(validated, nowIso = () => new Date().toISOString()) {
  const groups = new Map();

  for (const ch of validated) {
    const norm = normalizeName(ch.name);
    if (!norm) continue;
    if (!groups.has(norm)) groups.set(norm, []);
    groups.get(norm).push(ch);
  }

  const final = [];
  for (const [norm, group] of groups.entries()) {
    // Sort each group by latency (fastest first), dedupe identical URLs.
    const byLatency = [...group].sort((a, b) => a.latencyMs - b.latencyMs);
    const seen = new Set();
    const sources = [];
    for (const ch of byLatency) {
      if (!seen.has(ch.url)) {
        seen.add(ch.url);
        sources.push(ch.url);
      }
    }
    const primary = byLatency[0];
    const id = await sha256Hex16(norm);

    final.push({
      id,
      name: primary.name,
      logoUrl: primary.logo || '',
      streamUrl: primary.url,
      category: mapCategory(primary.category, primary.sourceId, primary.name),
      latencyMs: primary.latencyMs,
      language: primary.language,
      tier: primary.tier,
      sources,
      lastValidated: nowIso(),
    });
  }

  return final.sort((a, b) => a.name.localeCompare(b.name));
}

File: validate/dedupe.go
package main

import (
	"sort"
	"strings"
	"unicode"
)

type ChannelFinal struct {
	ID            string   `json:"id"`
	Name          string   `json:"name"`
	LogoURL       string   `json:"logoUrl"`
	StreamURL     string   `json:"streamUrl"`
	Category      string   `json:"category"`
	LatencyMs     int64    `json:"latencyMs,omitempty"`
	Language      string   `json:"language,omitempty"`
	Tier          string   `json:"tier,omitempty"`
	Sources       []string `json:"sources,omitempty"`
	LastValidated string   `json:"lastValidated,omitempty"`
	VideoCodec    string   `json:"videoCodec,omitempty"`
	AudioCodec    string   `json:"audioCodec,omitempty"`
}

func normalizeName(name string) string {
	cleaned := strings.ToLower(name)
	cleaned = strings.NewReplacer(
		"(hd)", "", "(fhd)", "", "(4k)", "", "(tv)", "",
		"[hd]", "", "[fhd]", "", "[4k]", "", "[tv]", "",
	).Replace(cleaned)

	var b strings.Builder
	for _, r := range cleaned {
		if isEmoji(r) || isZeroWidth(r) {
			continue
		}
		if unicode.IsLetter(r) || unicode.IsNumber(r) || r == ' ' || r == '-' {
			b.WriteRune(r)
		}
	}
	cleaned = b.String()
	cleaned = strings.Join(strings.Fields(cleaned), " ")
	return strings.TrimSpace(cleaned)
}

func pickBestRoutes(validated []ValidatedChannel) []ChannelFinal {
	groups := make(map[string][]ValidatedChannel)

	for _, ch := range validated {
		norm := normalizeName(ch.Name)
		if norm == "" {
			continue
		}
		groups[norm] = append(groups[norm], ch)
	}

	var result []ChannelFinal
	for _, group := range groups {
		sort.Slice(group, func(i, j int) bool {
			return group[i].LatencyMs < group[j].LatencyMs
		})

		norm := normalizeName(group[0].Name)

		seen := make(map[string]bool)
		var sources []string
		for _, ch := range group {
			if !seen[ch.URL] {
				seen[ch.URL] = true
				sources = append(sources, ch.URL)
			}
		}

		primary := group[0]
		id := generateID(norm)
		logo := normalizeLogoURL(primary.Logo)

		result = append(result, ChannelFinal{
			ID:            id,
			Name:          primary.Name,
			LogoURL:       logo,
			StreamURL:     primary.URL,
			Category:      primary.Category,
			LatencyMs:     primary.LatencyMs,
			Language:      primary.Language,
			Tier:          primary.Tier,
			Sources:       sources,
			LastValidated: primary.LastValidated,
			VideoCodec:    primary.VideoCodec,
		})
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].Name < result[j].Name
	})

	return result
}

func normalizeLogoURL(logo string) string {
	logo = strings.TrimSpace(logo)
	if logo == "" {
		return ""
	}
	if strings.Contains(logo, "imgur.com/") && !strings.HasPrefix(logo, "https://i.imgur.com/") {
		parts := strings.Split(logo, "imgur.com/")
		if len(parts) == 2 {
			return "https://i.imgur.com/" + parts[1]
		}
	}
	return logo
}
